### PR TITLE
ReverbSC: Method to change samplerate (#3401)

### LINKED ISSUE #3401: ReverbSC - Method to change sample rate
As mentioned here: https://github.com/LMMS/lmms/pull/3202#issuecomment-282132629

ReverbSC doesn't  have a method to change sample rate. This means the rate will stay the default when you export so if you choose another rate and/or oversampling the reverb will come out wrong.

### PR DESCRIPTION
**Pull request by PaulBatchelor**
This should fix #3401.

### CODE REVIEW

**Comment by zonkmachine** [comment]
Testing. Got crash but I don't know if this belongs in this or your threads/export issue.

```gdb
[New Thread 0x7fff9f7fe700 (LWP 16989)]
[Thread 0x7fff9effd700 (LWP 16987) exited]
*** Error in `/home/zonkmachine/builds/lmms/lmms/build/lmms': malloc(): smallbin double linked list corrupted: 0x0000000001d150f0 ***

Program received signal SIGABRT, Aborted.
0x00007ffff4100c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff4100c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff4104028 in __GI_abort () at abort.c:89
#2  0x00007ffff413d2a4 in __libc_message (do_abort=do_abort@entry=1, fmt=fmt@entry=0x7ffff424b6b0 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff4147ff7 in malloc_printerr (action=<optimized out>, str=0x7ffff424ba78 "malloc(): smallbin double linked list corrupted", 
    ptr=<optimized out>) at malloc.c:4996
#4  0x00007ffff414acf4 in _int_malloc (av=0x7ffff4488760 <main_arena>, bytes=16) at malloc.c:3359
#5  0x00007ffff414c6c0 in __GI___libc_malloc (bytes=16) at malloc.c:2891
#6  0x00007fffd4c15749 in sp_dcblock_create (p=0x7fffeaec2308) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/dcblock.c:19
#7  0x00007fffd4c116b3 in ReverbSCEffect::changeSampleRate (this=0x7fffeaec19e0) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSC.cpp:142
#8  0x00007fffd4c13a7c in ReverbSCControls::changeSampleRate (this=0x7fffeaec1f98)
    at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSCControls.cpp:66
#9  0x00007fffd4c15926 in ReverbSCControls::qt_static_metacall (_o=0x7fffeaec1fb8, _c=QMetaObject::InvokeMetaMethod, _id=1, _a=0x7fffa4000ca0)
    at /home/zonkmachine/builds/lmms/lmms/build/plugins/ReverbSC/moc_ReverbSCControls.cxx:51
#10 0x00007ffff6974c1e in QObject::event (this=0x7fffeaec1fb8, e=<optimized out>) at kernel/qobject.cpp:1194
#11 0x00007ffff70d2e2c in QApplicationPrivate::notify_helper (this=this@entry=0x9858f0, receiver=receiver@entry=0x7fffeaec1fb8, e=e@entry=0x7fffa4000cc0)
    at kernel/qapplication.cpp:4567
#12 0x00007ffff70d94a0 in QApplication::notify (this=0x976bc0, receiver=0x7fffeaec1fb8, e=0x7fffa4000cc0) at kernel/qapplication.cpp:4353
#13 0x00007ffff695c4dd in QCoreApplication::notifyInternal (this=0x976bc0, receiver=receiver@entry=0x7fffeaec1fb8, event=event@entry=0x7fffa4000cc0)
    at kernel/qcoreapplication.cpp:953
#14 0x00007ffff695fb3d in sendEvent (event=0x7fffa4000cc0, receiver=0x7fffeaec1fb8) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:231
#15 QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0x985ae0)
    at kernel/qcoreapplication.cpp:1577
```

**Comment by PaulBatchelor** [comment]
Oh bummer. Yeah, I was afraid something like this would happen. It doesn't look like malloc can be safely called here. I can try adding a mutex here to see if that fixes things.

**Comment by PaulBatchelor** [comment]
Just pushed changes. @zonkmachine can you try it again?

**Comment by zonkmachine** [comment]
I have seen no more crashes so this seem fixed.
Next, oversampling! There is a difference in the output between oversampling settings. Here are 3 different reverb units compared with no oversampling vs. 8 times, ReverbSC on the bottom.
Thought. Could this be because of how ReverbSC use randomness so you get the equivalent to noise reduction here?

reverbinterpolation

**Comment by PaulBatchelor** [comment]
Hmmm... so this reverb algorithm (which I didn't write), uses variable delay lines which drift. It would appear that the delay line times are always based on a samplerate of 44.1kHz:

https://github.com/LMMS/lmms/blob/master/plugins/ReverbSC/revsc.c#L36

These constants never seem to change. I just assumed they did.

Going to have to fix this on the Soundpipe branch. Great catch!

Also, the sample rates you are using get quite large. It would appear that I may need to change the type for the samplerate variable ( currently a signed int): https://github.com/LMMS/lmms/blob/master/plugins/ReverbSC/base.h#L28

Even more good improvements for Soundpipe!

**Comment by zonkmachine** [comment]
Cool!

Also I spotted this random message during **make:** `/home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSC.h:0: Note: No relevant classes found. No output generated.`
I think this is not a new message though as I got it from PR #3404 and also again when recompiling with this PR cherry-picked on top.

**Comment by PaulBatchelor** [comment]
Oh yeah... I've been ignoring that up to this point. google tells me qt gives that warning. My guess is that it is some tweaking in the cmake file.

**Comment by PaulBatchelor** [comment]
Making some tweaks to the CMake file in an effort to remove warnings. Also pushing a change to make the samplerate a uint32_t instead of just an int to prevent overflow.

**Comment by PaulBatchelor** [comment]
Also, misunderstood your query about the ReverbSC. Thought those were *all* reverbSC with different resampling rates. Didn't realize you were comparing them.

I still can't explain what causes the change in waveform to happen when oversampling is used. More information can be extracted from a spectrogram. I recommend using something like sox to generate it.

**Comment by zonkmachine** [comment]
> Making some tweaks to the CMake file in an effort to remove warnings.

Warnings gone.

> More information can be extracted from a spectrogram. I recommend using something like sox to generate it.

I tried sox but the test project I used came out looking a bit too homogeneous. Nothing there to really catch ones eye. Do you have any special sample that makes a good test case and/or a better command than just running `sox test1.wav -n spectrogram` ?

**Comment by PaulBatchelor** [comment]
Send an impulse through the reverb unit to get the impulse response. All the information about the reverb characteristics can be seen in the reverb tail.
A good impulse to use is a short white noise burst (1-5ms) or single sample tick. Pretty easy things to make in sporth:

A single sample tick:

'''
echo "tick" | sporth -o tick.wav -d 1s
'''

A 4ms broadband noise click:
'''
echo "tick 0.004 tgate 1.0 noise *" | sporth -o tick.wav -d 1s
'''

**Comment by PaulBatchelor** [comment]
Send an impulse through the reverb unit to get the impulse response. All the information about the reverb characteristics can be seen in the reverb tail.
A good impulse to use is a short white noise burst (1-5ms) or single sample tick. Pretty easy things to make in sporth:

A single sample tick:

```
echo "tick" | sporth -o tick.wav -d 1s
```

A 4ms broadband noise click:
```
echo "tick 0.004 tgate 1.0 noise *" | sporth -o tick.wav -d 1s
```

**Comment by zonkmachine** [comment]
tick + long reverb -> 44100 x1 oversampling

spectrogram1

tick + long reverb -> 44100 x8 oversampling

spectrogram2

Samples: [soundfiles.zip](https://github.com/LMMS/lmms/files/825894/soundfiles.zip)

**Comment by zonkmachine** [comment]
^updated

**Comment by PaulBatchelor** [comment]
Interesting! 

There is some content at around DC that goes away with the 8x oversampling. This could be the noise reduction you were talking about. This reverb produces a lot of DC offset, so I'm using some dc blockers here. My intuition is telling me that the oversampling could probably be helping out the DC blockers do their job in filtering out the low low end junk, but take that with a grain of salt.

**Comment by zonkmachine** [comment]
OT but I think you have a test in `dcblock.c` that is never going to evaluate as true.
https://github.com/LMMS/lmms/blob/master/plugins/ReverbSC/dcblock.c#L33:L35

**Comment by PaulBatchelor** [comment]
Heh. You would be right. Thanks!

**Comment by zonkmachine** [comment]
Interesting. I poked this one a bit and here's what I've got.
`p->gain = 0.99;`
Increase this to 0.9999 and the problems change. I think you need this one depending on the oversampling factor to get the same precise filtering on all export cases so the fix would be something in the vein of (In C++):

```cpp
p->gain = pow( 0.99, ( 1.0f / Engine::mixer->sampleRateMultiplier() ) );
```

I think this one works but I don't know what to look for when comparing spectrogram from different oversampling settings.

over1


over8fixed

**Comment by zonkmachine** [comment]
Yes, it looks like a fix. Here's the spectrogram for the same sample without the fix and oversampling x 8. 
They look identical in the higher frequencies.

p->gain = 0.99
oversampling x 8
x8

**Comment by zonkmachine** [comment]
Fix here: https://gist.github.com/zonkmachine/08bec433cda8aede4e9e3dc636dbdfba

**Comment by PaulBatchelor** [comment]
Sure, okay.

**Comment by zonkmachine** [comment]
I got another crash, similar to the one before. Apart from the fix above I had modified the code to `p->gain = 0.9999` , as one does, and the crash is in sp_dcblock but ahead of the modified part. The crash came near the end of the render and I believe the settings were highest quality on everything. 192kHz / Sinc Best / x8.

```
Core was generated by `./lmms'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  _int_malloc (av=0x7f25d82b6760 <main_arena>, bytes=16) at malloc.c:3389
3389	malloc.c: No such file or directory.
(gdb) bt
#0  _int_malloc (av=0x7f25d82b6760 <main_arena>, bytes=16) at malloc.c:3389
#1  0x00007f25d7f7a6c0 in __GI___libc_malloc (bytes=16) at malloc.c:2891
#2  0x00007f25b8547afb in sp_dcblock_create (p=0x7f25ceccad48) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/dcblock.c:20
#3  0x00007f25b854395b in ReverbSCEffect::changeSampleRate (this=0x7f25cecca420) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSC.cpp:143
#4  0x00007f25b8545db4 in ReverbSCControls::changeSampleRate (this=0x7f25cecca9d8)
    at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSCControls.cpp:66
#5  0x00007f25b8547d18 in ReverbSCControls::qt_static_metacall (_o=0x7f25cecca9f8, _c=QMetaObject::InvokeMetaMethod, _id=1, _a=0x7f2578004f70)
    at /home/zonkmachine/builds/lmms/lmms/build/plugins/ReverbSC/moc_ReverbSCControls.cxx:51
```

**Comment by PaulBatchelor** [comment]
It's crashing at malloc again, which is what I thought the mutex would fix.

**Comment by zonkmachine** [comment]
changeSampleRate() is called once when you start exporting but twice when export is done. Could this be a problem if the calls are made close to each other?

**Comment by PaulBatchelor** [comment]
> changeSampleRate() is called once when you start exporting but twice when export is done. Could this be a problem if the calls are made close to each other?

It wouldn't surprise me if there was an issue related to this. ReverbSC could very well be the only plugin using malloc in changeSampleRate, which does not bode well for highly threaded applications like LMMS.

And why is it called twice anyways?

**Comment by zonkmachine** [comment]
> And why is it called twice anyways?

As to 'why' in particular, I don't know but it happens here:

In the end of ProjectRenderer::run()[ these calls ](https://github.com/LMMS/lmms/blob/master/src/core/ProjectRenderer.cpp#L200:L201)are made to two functions [Mixer::restoreAudioDevice()](https://github.com/LMMS/lmms/blob/master/src/core/Mixer.cpp#L614:L627) and [Mixer::changeQuality()](https://github.com/LMMS/lmms/blob/master/src/core/Mixer.cpp#L542:L554) that both each emits `sampleRateChanged()` .

I put some printf() statements in these functions and in `ReverbSCEffect::changeSampleRate()` to see in what order things happended and the results was either

```
Mixer::restoreAudioDevice()
changeSampleRate()
Mixer::changeQuality()
changeSampleRate()
```
or

```
Mixer::restoreAudioDevice()
Mixer::changeQuality()
changeSampleRate()
changeSampleRate()
```

**Comment by zonkmachine** [comment]
Some gdb multi-thread output. 

Backtraces as gist: 
[(gdb) bt](https://gist.github.com/zonkmachine/af069af4c2d79ca3327834641628de24#file-backtraces-txt-L12)
[(gdb) bt full](https://gist.github.com/zonkmachine/af069af4c2d79ca3327834641628de24#file-backtraces-txt-L36)
[(gdb) thread apply all bt full](https://gist.github.com/zonkmachine/af069af4c2d79ca3327834641628de24#file-backtraces-txt-L133)


Two threads accessing ReverbSC. 
```gdb
(gdb)  thread apply all bt

Thread 2 (Thread 0x7f7e9ee4a7c0 (LWP 31733)):
#0  sp_dcblock_create (p=0x181c800000008) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/dcblock.c:19
#1  0x00007f7e85edb9d5 in ReverbSCEffect::changeSampleRate (this=0x7f7e91cfa120) at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSC.cpp:150
#2  0x00007f7e85edde64 in ReverbSCControls::changeSampleRate (this=0x7f7e91cfa6d8)
    at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSCControls.cpp:66
#3  0x00007f7e85edfdc8 in ReverbSCControls::qt_static_metacall (_o=0x7f7e91cfa6f8, _c=QMetaObject::InvokeMetaMethod, _id=1, _a=0x7f7e540045d0)
    at /home/zonkmachine/builds/lmms/lmms/build/plugins/ReverbSC/moc_ReverbSCControls.cxx:51
#4  0x00007f7e9d7fdc1e in QObject::event (this=0x7f7e91cfa6f8, e=<optimized out>) at kernel/qobject.cpp:1194
#5  0x00007f7e9d7e54dd in QCoreApplication::notifyInternal (this=0x1312bc0, receiver=receiver@entry=0x7f7e91cfa6f8, event=event@entry=0x7f7e540045f0)
    at kernel/qcoreapplication.cpp:953
#6  0x00007f7e9d7e8b3d in sendEvent (event=0x7f7e540045f0, receiver=0x7f7e91cfa6f8) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:231
#7  QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0x1321a60)
    at kernel/qcoreapplication.cpp:1577
#8  0x00007f7e9d7e8fe3 in QCoreApplication::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0)
    at kernel/qcoreapplication.cpp:1470
#9  0x00007f7e9d812f83 in sendPostedEvents () at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:236
#10 postEventSourceDispatch (s=0x1323ed0) at kernel/qeventdispatcher_glib.cpp:287
#11 0x00007f7e9a810e04 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#12 0x00007f7e9a811048 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#13 0x00007f7e9a8110ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#14 0x00007f7e9d8127a1 in QEventDispatcherGlib::processEvents (this=0x1323b90, flags=...) at kernel/qeventdispatcher_glib.cpp:434
#15 0x00007f7e9d7e40af in QEventLoop::processEvents (this=this@entry=0x7ffe1241dc70, flags=...) at kernel/qeventloop.cpp:149
#16 0x00007f7e9d7e43a5 in QEventLoop::exec (this=this@entry=0x7ffe1241dc70, flags=...) at kernel/qeventloop.cpp:204
#17 0x00007f7e9d7e9b79 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1225
#18 0x00000000004dfaa7 in main (argc=5, argv=0x7ffe1241e068) at /home/zonkmachine/builds/lmms/lmms/src/core/main.cpp:901

Thread 1 (Thread 0x7f7e5d78b700 (LWP 31741)):
#0  0x00007f7e85edf79a in sp_revsc_compute (sp=0x1596250, p=0x1522ff0, in1=0x7f7e5d78ab40, in2=0x7f7e5d78ab44, out1=0x7f7e5d78ab20, out2=0x7f7e5d78ab24)
    at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/revsc.c:201
#1  0x00007f7e85edb745 in ReverbSCEffect::processAudioBuffer (this=0x7f7e91cfa120, buf=0x7f7e91c980a0, frames=256)
    at /home/zonkmachine/builds/lmms/lmms/plugins/ReverbSC/ReverbSC.cpp:118
#2  0x0000000000507c43 in EffectChain::processAudioBuffer (this=0x156a720, _buf=0x7f7e91c980a0, _frames=256, hasInputNoise=false)
    at /home/zonkmachine/builds/lmms/lmms/src/core/EffectChain.cpp:217
#3  0x000000000056977f in AudioPort::processEffects (this=0x7f7e91ca2370) at /home/zonkmachine/builds/lmms/lmms/src/core/audio/AudioPort.cpp:99
#4  0x000000000056a151 in AudioPort::doProcessing (this=0x7f7e91ca2370) at /home/zonkmachine/builds/lmms/lmms/src/core/audio/AudioPort.cpp:223
#5  0x0000000000517651 in ThreadableJob::process (this=0x7f7e91ca2370) at /home/zonkmachine/builds/lmms/lmms/include/ThreadableJob.h:74
#6  0x000000000052eb5b in MixerWorkerThread::JobQueue::run (this=0x94a420 <MixerWorkerThread::globalJobQueue>)
    at /home/zonkmachine/builds/lmms/lmms/src/core/MixerWorkerThread.cpp:75
#7  0x000000000052edb5 in MixerWorkerThread::startAndWaitForJobs () at /home/zonkmachine/builds/lmms/lmms/src/core/MixerWorkerThread.cpp:148
#8  0x0000000000529a1b in Mixer::renderNextBuffer (this=0x13459c0) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:462
#9  0x000000000052b3d7 in Mixer::fifoWriter::run (this=0x7f7e54003250) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:1102
#10 0x00007f7e9d6e332f in QThreadPrivate::start (arg=0x7f7e54003250) at thread/qthread_unix.cpp:349
#11 0x00007f7e9ea58184 in start_thread (arg=0x7f7e5d78b700) at pthread_create.c:312
#12 0x00007f7e9b02237d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb)
```

**Comment by PaulBatchelor** [comment]
You'll have to show me how to do that trick. I have no experience debugging threads.

Looks like the audio is still processing audio when the sample rate is changed, which shouldn't be happening. I fixed a similar problem in #3404, so I imagine this is in the same family of bug.

**Comment by zonkmachine** [comment]
> You'll have to show me how to do that trick. I have no experience debugging threads.

Me neither. This is actually the fist time I've used the command to show all threads that it's actually give me something useful back and now you know as much as I do about it  ;-)

**Comment by PaulBatchelor** [comment]
@zonkmachine ``Mixer::renderNextBuffer()`` shouldn't be running while ``changeSampleRate()`` is being called. 

It's also a little ambiguous to me where that buffer is going. Is it going to my speakers, or is it going to an audio file. When I was fixing the CLI renderer, I realized that JACK was running while the file was being written to disk, and this was causing some race conditions.

**Comment by zonkmachine** [comment]
If I turn down the decay on the FX unit I no longer get a crash (~100 exports, no crash). I see the crash under both Alsa and SDL.

I think the general idea is to push out the next release candidate, RC3, pretty soon. If the ultimate fix for this is longer down the road I wonder if it would be a good idea to just push/merge what you've got so far? (use whatever you can from the oversampling fix. I'm not git competent enough to push anything... :) ). For me the crashes are pretty infrequent and I'd prefer occasional crashes to discovering that the track sounds different on export. 
Unfortunately I don't think there is anything more I can do here other than test whatever you throw at me.

**Comment by PaulBatchelor** [comment]
> If the ultimate fix for this is longer down the road I wonder if it would be a good idea to just push/merge what you've got so far? 

Well I have no such fixes yet (the only thing close to it is the other PR I've made). Once the exporter for LMMS is fixed (and yes, it's very broken), this problem will go away.

**Comment by PaulBatchelor** [comment]
@zonkmachine I suggest you merge the code I have pushed already, as this theoretical bugfix is beyond the scope of this PR.

**Comment by Umcaruje** [comment]
Changed the PR to address `stable-1.2`

**Comment by zonkmachine** [comment]
So, merge this!

I suggested a fix https://github.com/LMMS/lmms/pull/3403#issuecomment-285343128 here to [some issues with this PR](https://github.com/LMMS/lmms/pull/3403#issuecomment-284210178) that was more of a "something like this would work" than a fix, even though it works.
Should this be incorporated in this PR or pulled separately? I've never committed anything to someone else's PR.

**Comment by tresf** [comment]
>  I've never committed anything to someone else's PR.

Me neither until [here](https://github.com/LMMS/lmms/pull/3489/commits) documented [here](https://github.com/tresf/lmms/wiki/Bollocks).  Works fine.  You just reference the originating branch as if it lives here in upstream.  Works perfectly.

**Comment by zonkmachine** [comment]
> Me neither until here documented here. Works fine. You just reference the originating branch as if it lives here in upstream. Works perfectly.

Sorry, I didn't get the instructions.

**Comment by zonkmachine** [comment]
@tresf Bump...

**Comment by tresf** [comment]
> Sorry, I didn't get the instructions. [...] Bump... 

I assume the bump is for the instructions?  Here's my own interpretation of them... https://github.com/tresf/lmms/wiki/Bollocks

**Comment by tresf** [comment]
Sorry... those instructions aren't very good...

Basically, you just reference the branch as if it resides at `lmms/lmms` and you can commit directly to it prior to merge.

**Comment by zonkmachine** [comment]
> Basically, you just reference the branch as if it resides at lmms/lmms and you can commit directly to it prior to merge.

Sorry, I would need step by step instructions for this.

**Comment by tresf** [comment]
```bash
git remote add paulbatchelor https://github.com/paulbatchelor/lmms
git fetch paulbatchelor
# normally, we'd just checkout the branch, but paul made the mistake of using "master", so we need to specify the new master
git checkout -b master paulbatchelor master
# make any changes
git push paulbatchelor master
```

If you get access denied, this may be a security feature due to the fact he opened a PR with his master branch.  If that happens, just close this PR and open a new one with all of the changes applied.  They can be cherry-picked by hash onto a new branch, I'd be happy to set this up for you if needed.

**Comment by zonkmachine** [comment]
```
$ git checkout -b master paulbatchelor master
fatal: A branch named 'master' already exists.
```

**Comment by tresf** [comment]
@zonkmachine amended (changed local name to `master2` put `/` between `paulbatchelor` and `master`), can you try it again?

**Comment by zonkmachine** [comment]
> can you try it again?

Yes, but something went wrong. I'll try and revert this.

**Comment by zonkmachine** [comment]
I don't know what went wrong. I reset master2 to a backup and try to push over @PaulBatchelor 's now borked master. ( so sorry! )

```
$ git push -f PaulBatchelor  master
Username for 'https://github.com': zonkmachine
Password for 'https://zonkmachine@github.com': 
Everything up-to-date

```

Up to date, really?

```

zonkmachine@zonkmachine:~/builds/lmms/lmms$ git status 
On branch master2
Your branch and 'PaulBatchelor/master' have diverged,
and have 4 and 58 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

```

**Comment by zonkmachine** [comment]
` git push paulbatchelor master`

This seem to have pushed my master to Paul's master.

**Comment by zonkmachine** [comment]
Unborked with `git push -f PaulBatchelor  master2:master`

**Comment by zonkmachine** [comment]
Nailed it! :man_dancing:

File: plugins/ReverbSC/dcblock.c [hunk 1]
```diff
@@ -26,11 +27,11 @@ int sp_dcblock_destroy(sp_dcblock **p)
     return SP_OK;
 }
 
-int sp_dcblock_init(sp_data *sp, sp_dcblock *p)
+int sp_dcblock_init(sp_data *sp, sp_dcblock *p, int oversampling )
 {
     p->outputs = 0.0;
     p->inputs = 0.0;
-    p->gain = 0.99;
+    p->gain = pow( 0.99, 1.0f / oversampling );
```
Comment (zonkmachine) [inline plugins/ReverbSC/dcblock.c]: The idea here is that the oversampling causes the net gain to vary which changes the frequency response of the DC block. Now we always get a gain of 0.99 .

**Comment by zonkmachine** [comment]
There still is the issue of occasional crashes on export to which there currently is no fix. I think this should be merged anyway. It's a minor issue to the variable frequency response on export with oversampling.

**Comment by zonkmachine** [comment]
I think we should just merge this and open a separate issue for the export crashes.

**Comment by zonkmachine** [comment]
What's the deal with rebasing PR's from multiple persons? Do I squash to one commit as usually?

**Comment by tresf** [comment]
You can or it can be done through the web interface.  In my experience both commiters get credit and both are listed on the single commit.

**Pull request merged**